From 5185276590e2099c55292fec2906e8d0f8a5bc16 Mon Sep 17 00:00:00 2001
From: Konstantin Khlebnikov <khlebnikov@openvz.org>
Date: Mon, 17 Mar 2008 15:36:28 +0300
Subject: [PATCH 35/48] Change sys_setsid() and set_special_pids() to work with struct pid.

Backport mainstream patch from Oleg Nesterov <oleg@tv-sign.ru> .
---
 include/linux/sched.h |    6 ++----
 init/main.c           |    2 +-
 kernel/exit.c         |   21 +++++++++++----------
 kernel/sys.c          |   11 ++++++-----
 4 files changed, 20 insertions(+), 20 deletions(-)

Index: kernel/include/linux/sched.h
===================================================================
--- kernel.orig/include/linux/sched.h	2008-11-24 15:57:08.000000000 +0100
+++ kernel/include/linux/sched.h	2008-11-24 15:58:47.000000000 +0100
@@ -1665,8 +1665,6 @@
 extern struct task_struct *find_task_by_pid_ns(pid_t nr,
 		struct pid_namespace *ns);
 
-extern void __set_special_pids(pid_t session, pid_t pgrp);
-
 /* per-UID process charging. */
 extern struct user_struct * alloc_uid(struct user_namespace *, uid_t);
 static inline struct user_struct *get_uid(struct user_struct *u)
@@ -1678,8 +1676,8 @@
 extern void switch_uid(struct user_struct *);
 extern void release_uids(struct user_namespace *ns);
 extern int set_user(uid_t uid, int dumpclear);
-extern void set_special_pids(pid_t session, pid_t pgrp);
-extern void __set_special_pids(pid_t session, pid_t pgrp);
+extern void set_special_pids(struct pid *pid);
+extern void __set_special_pids(struct pid *pid);
 
 #include <asm/current.h>
 
Index: kernel/init/main.c
===================================================================
--- kernel.orig/init/main.c	2008-11-24 15:57:27.000000000 +0100
+++ kernel/init/main.c	2008-11-24 15:58:47.000000000 +0100
@@ -858,7 +858,7 @@
 	 */
 	init_pid_ns.child_reaper = current;
 
-	__set_special_pids(1, 1);
+	__set_special_pids(&init_struct_pid);
 	cad_pid = task_pid(current);
 
 	smp_prepare_cpus(max_cpus);
Index: kernel/kernel/exit.c
===================================================================
--- kernel.orig/kernel/exit.c	2008-11-24 15:57:08.000000000 +0100
+++ kernel/kernel/exit.c	2008-11-24 15:58:47.000000000 +0100
@@ -310,26 +310,27 @@
 	switch_uid(INIT_USER);
 }
 
-void __set_special_pids(pid_t session, pid_t pgrp)
+void __set_special_pids(struct pid *pid)
 {
 	struct task_struct *curr = current->group_leader;
+	pid_t nr = pid_nr(pid);
 
-	if (task_session_nr(curr) != session) {
+	if (task_session(curr) != pid) {
 		detach_pid(curr, PIDTYPE_SID);
-		set_task_session(curr, session);
-		attach_pid(curr, PIDTYPE_SID, find_pid(session));
+		attach_pid(curr, PIDTYPE_SID, pid);
+		set_task_session(curr, nr);
 	}
-	if (task_pgrp_nr(curr) != pgrp) {
+	if (task_pgrp(curr) != pid) {
 		detach_pid(curr, PIDTYPE_PGID);
-		set_task_pgrp(curr, pgrp);
-		attach_pid(curr, PIDTYPE_PGID, find_pid_ns(pgrp, &init_pid_ns));
+		attach_pid(curr, PIDTYPE_PGID, pid);
+		set_task_pgrp(curr, nr);
 	}
 }
 
-void set_special_pids(pid_t session, pid_t pgrp)
+void set_special_pids(struct pid *pid)
 {
 	write_lock_irq(&tasklist_lock);
-	__set_special_pids(session, pgrp);
+	__set_special_pids(pid);
 	write_unlock_irq(&tasklist_lock);
 }
 EXPORT_SYMBOL(set_special_pids);
@@ -401,7 +402,7 @@
 	 */
 	current->flags |= PF_NOFREEZE;
 
-	set_special_pids(1, 1);
+	set_special_pids(&init_struct_pid);
 	proc_clear_tty(current);
 
 	/* Block and flush all signals */
Index: kernel/kernel/sys.c
===================================================================
--- kernel.orig/kernel/sys.c	2008-11-24 15:47:46.000000000 +0100
+++ kernel/kernel/sys.c	2008-11-24 15:58:47.000000000 +0100
@@ -1192,6 +1192,7 @@
 {
 	struct task_struct *group_leader = current->group_leader;
 	pid_t session;
+	struct pid *sid;
 	int err = -EPERM;
 
 	write_lock_irq(&tasklist_lock);
@@ -1200,7 +1201,8 @@
 	if (group_leader->signal->leader)
 		goto out;
 
-	session = group_leader->pid;
+	sid = task_pid(group_leader);
+	session = pid_vnr(sid);
 	/* Fail if a process group id already exists that equals the
 	 * proposed session id.
 	 *
@@ -1208,19 +1210,18 @@
 	 * session id and so the check will always fail and make it so
 	 * init cannot successfully call setsid.
 	 */
-	if (session > 1 && find_task_by_pid_type_ns(PIDTYPE_PGID,
-				session, &init_pid_ns))
+	if (session > 1 && pid_task(sid, PIDTYPE_PGID))
 		goto out;
 
 	group_leader->signal->leader = 1;
-	__set_special_pids(session, session);
+	__set_special_pids(sid);
 
 	spin_lock(&group_leader->sighand->siglock);
 	group_leader->signal->tty = NULL;
 	group_leader->signal->tty_old_pgrp = 0;
 	spin_unlock(&group_leader->sighand->siglock);
 
-	err = task_pgrp_vnr(group_leader);
+	err = session;
 out:
 	write_unlock_irq(&tasklist_lock);
 	return err;
